apply snapshot in background apply fiber if required

mmaslankaprv opened the pull request:
The background apply fiber is used in state machine manger to catch up
the state machines which are lagging (f.e. state machine apply method
thrown an exception). The background apply fiber was missing the
snapshot application. This lead to a rare situation in which a snapshot
delivery while the state machine was being caught up in background apply
might have lead to situation in which a reader can not continue as the
next offset was not updated and below the raft start offset.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
r/stm_manager: added log line when starting stm manager with stm offset

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/stm_manager: apply snapshot in background apply fiber if required

The background apply fiber is used in state machine manger to catch up
the state machines which are lagging (f.e. state machine apply method
thrown an exception). The background apply fiber was missing the
snapshot application. This lead to a rare situation in which a snapshot
delivery while the state machine was being caught up in background apply
might have lead to situation in which a reader can not continue as the
next offset was not updated and below the raft start offset.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/tests: added test case for snapshot apply in stm_mgr background apply

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -218,6 +218,15 @@ ss::future<> state_machine_manager::stop() {
+    std::ranges::copy(
+      std::views::values(_machines), std::back_inserter(all_stms));
+    return all_stms;
+}
```
Is it critical to copy it here, so that we snapshot the set of machines? If not, wouldn't it be neater for `do_apply_raft_snapshot` to accept a templated range argument instead?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--7d611881-6788-4dc4-b324-bc4d8a34c6d6-->
<!--TEST RESULTS-->
<!--32a8e6f4-226f-43d8-9f20-15c93869eefe-->
<!--TEST RESULTS-->
<!--db569c7e-29d6-43e0-ad3c-be3192fe7c3a-->
<!--TEST RESULTS-->
<!--0e0a702d-dad4-4853-8b7f-0f881a7c2f52-->
<!--TEST RESULTS-->
<!--c7f7fc90-d368-47ae-b9d1-890428d467bc-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64490">build#64490</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64490#01963529-3694-4ac3-a0b5-8b5efe23f22a | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64490#01963529-3695-4966-9664-95ff197b149b | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64601">build#64601</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64601#01963d77-fc9a-42d6-b136-5a73f14ba014 | FLAKY | 1/2 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64601#01963d77-fc98-4957-8d3e-d34e7437ba45 | FLAKY | 1/3 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.TIRED_STORAGE.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/64601#01963d77-fc9a-42d6-b136-5a73f14ba014 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64601#01963d77-fc99-4449-9d0b-ebacc823c432 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64630">build#64630</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64630#01963fda-2232-4443-8778-85342a6d0f54 | FLAKY | 1/2 |
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64630#01963fda-2232-4443-8778-85342a6d0f54 | FLAKY | 1/2 |
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64630#01964035-39b7-470f-8dc6-7a7b6dd4ec32 | FLAKY | 20/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64630#01964031-bc71-4b9e-9005-9f4469163469 | FLAKY | 16/21 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/64630#01964035-39b6-454f-8873-a71093ae192d | FLAKY | 16/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64630#01964035-39b5-46ad-aa0e-2ec82c0d2bbb | FLAKY | 11/21 |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=True.allow_gaps_cluster_level=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64630#01964035-39b4-441c-9943-01eb15abcad4 | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64669">build#64669</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/64669#019643c4-9b5a-4777-b8c6-09a3c60bced8 | FAIL | 0/1 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64669#019643c4-9b56-4fd3-baf9-c1bedd15c11e | FLAKY | 16/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64669#019643c4-9b5a-4777-b8c6-09a3c60bced8 | FAIL | 0/1 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all.use_transactions=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64669#019643c1-ee17-4fae-a8b7-f495f8bd1ca8 | FLAKY | 19/21 |
| rptest.tests.write_caching_fi_test.WriteCachingFailureInjectionTest.test_unavoidable_data_loss | ducktape | https://buildkite.com/redpanda/redpanda/builds/64669#019643c4-9b5a-4777-b8c6-09a3c60bced8 | FAIL | 0/1 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64669#01964369-cd2c-43ba-b158-b7fd9a762742 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64687">build#64687</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64687#019644ce-210f-44fa-9380-7788126fab6f | FLAKY         | 8/21     |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64687#019644b9-b2e5-424a-8e0c-060d4dc34a8a | FLAKY         | 19/21    |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=off.test_mode=TestMode.FAST_MOVES.cleanup_policy=compact       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64687#019644b9-b2e1-456b-84a0-26387886c848 | FLAKY         | 20/21    |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64687#019644ce-210f-44fa-9380-7788126fab6f | FLAKY         | 13/21    |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64687#019644ce-210f-44fa-9380-7788126fab6f | FLAKY         | 19/21    |
</details>
<!--/body-->

nvartolomei commented on src/v/raft/tests/persisted_stm_test.cc:
```diff
@@ -438,7 +446,7 @@ struct persisted_stm_test_fixture : state_machine_fixture {
                 });
           });
-
+        fmt::print(">>>> snapshot offset: {}\n", snapshot_offset);
```
leftover or we vlog is not liked int he raft tests?

mmaslankaprv replied:
leftover

nvartolomei reviewed (commented):
Discussed offline with @mmaslankaprv that there might be possible to have a nicer fix. 

The key observation is that in the "foreground" apply snapshot we do wait on mutexes for all STMs, including the ones stuck with background apply so the application never succeeds for any STMs! (Interestingly we didn't notice this in the incident. I would have guessed much more would have been broken if STMs are stuck).

With that observation, we can just bail out of the background apply loop when _next is below start offset because we know that there is a pending snapshot apply waiting on our lock.

nvartolomei reviewed (dismissed):
lgtm

from @mmaslankaprv 

> i actually need the snapshot application in background apply as the `state_machine_manager::_next` may already be greater than the last included snapshot index while one of the state machines _next is smaller. In this case the snapshot would not be applied

mmaslankaprv commented:
/cdt

rockwotj commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -444,6 +462,34 @@ void state_machine_manager::maybe_start_background_apply(
+        if (entry->stm->next() < _raft->start_offset()) {
+            auto snapshot = co_await _raft->open_snapshot();
+            if (!snapshot) {
+                co_return;
```
why do we bail if there is no snapshot? Should this be an error log at least?

nvartolomei replied:
wow 😶‍🌫️

mmaslankaprv committed:
r/stm_manager: fixed bug in stm manager leading to unnecessary reads

The state machine manager should select the max offset from all the
available so the lagging state machines can be caught up in the
background apply. Previously the stm manager selected the smallest
offset leading to catching up the state machines in foreground and
trying to call apply on the state machines which are already up to date.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/persisted_stm_test: improve persisted stm test coverage

Previously the test state machines were using the same snapshot making
the tests not meaningful. All the state machines started with the same
offset.

Added tests verifying adding a state machine later in the raft group
lifecycle f.e after the local or raft snapshot is taken.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

rockwotj commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -444,6 +460,40 @@ void state_machine_manager::maybe_start_background_apply(
+            auto snapshot = co_await _raft->open_snapshot();
+            if (!snapshot) {
+                vlog(
+                  _log.warn,
```
is there a circumstance where we expect this to happen? I am wondering why warn over error. I guess it will retry every 100ms, which should be fine?

mmaslankaprv replied:
Only when partition is deleted

rockwotj replied:
Hrm but should the STMs be shutdown there? I guess a warning is fine.

bashtanov commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -174,6 +174,14 @@ state_machine_manager::state_machine_manager(
+      std::ranges::max_element(_machines, [](auto& lhs, auto& rhs) {
+          return lhs.second->stm->last_applied_offset()
+                 < rhs.second->stm->last_applied_offset();
+      })->second->stm->last_applied_offset());
```
```
    auto offsets = _machines | std::views::transform([](const auto& entry) {
                       return entry.second->stm->last_applied_offset();
                   });
    return model::next_offset(*std::ranges::max_element(offsets));
```
all C++20

bashtanov commented on src/v/raft/tests/persisted_stm_test.cc:
```diff
@@ -766,3 +774,85 @@ TEST_F_CORO(state_machine_fixture, test_concurrent_apply_and_snapshot) {
+    for (auto [id, stm] : node_stms) {
+        if (stm->last_applied() == offset) {
+            snapshot
+              = co_await node(id.id()).raft()->stm_manager()->take_snapshot();
```
break here to conform to the comment that says we take a snapshot on _one_ node?

bashtanov commented on src/v/raft/tests/persisted_stm_test.cc:
```diff
@@ -766,3 +774,85 @@ TEST_F_CORO(state_machine_fixture, test_concurrent_apply_and_snapshot) {
+ * up.
+ */
+TEST_F_CORO(persisted_stm_test_fixture, test_snapshot_in_background_apply) {
+    ss::shared_ptr<other_persisted_kv> other_stm;
```
IIUC this variable holds the second STM for node 2. Do we need it?

bashtanov commented on src/v/raft/tests/persisted_stm_test.cc:
```diff
@@ -766,3 +774,85 @@ TEST_F_CORO(state_machine_fixture, test_concurrent_apply_and_snapshot) {
 }
+
+/**
+ * Tests the scenario in which an STM is added to the partition after it was
+ * already alive and Raft snapshot was taken on the partition.
```
I'm not sure if I understand the scenario. The code adds both STMs at the same time before taking any raft snapshots.

mmaslankaprv committed:
r/stm_manager: made last_applied a private method

Last applied returned a value based on `_next` which is a state machine
manager internal state. We should not leak it outside the
implementation.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

rockwotj reviewed (dismissed):
LGTM

bashtanov commented on src/v/raft/tests/persisted_stm_test.cc:
```diff
@@ -766,3 +779,215 @@ TEST_F_CORO(state_machine_fixture, test_concurrent_apply_and_snapshot) {
+    co_await wait_for_apply();
+    auto leader_id = co_await wait_for_leader(10s);
+    for (const auto& [_, stm] : node_stms) {
+        ASSERT_EQ_CORO(stm->state, expected);
+        ASSERT_EQ_CORO(stm->last_operation, other_stm->last_operation);
+        ASSERT_LE_CORO(
+          stm->apply_count, node(leader_id).raft()->committed_offset()());
```
any of these failed before the fix?

mmaslankaprv replied:
yes

bashtanov replied:
so which one out of curiosity?

mmaslankaprv replied:
test_adding_state_machine_no_raft_snapshot
test_adding_state_machine

mmaslankaprv committed:
r/tests: adjust persisted stm test timeout

Now that we have more test the test can take longer to finish.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64630-->
<!--6d2a3b51-1caf-4866-bdda-748ce7920280-->
<!--504588cd-ca60-4f5a-9bf4-fa1d4f6f9372-->
<!--c2078573-986e-440f-9e38-943deb45a9b8-->
<!--41d24323-21ba-497b-8293-382c6d4d3949-->
<!--65c874cb-cec1-44e6-8687-7f0707ba9776-->
<!--title-->
## Retry command for [Build#64630](https://buildkite.com/redpanda/redpanda/builds/64630)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
tests/rptest/tests/tiered_storage_pause_test.py::TestTieredStoragePause.test_safe_pause_resume@{"allow_gaps_cluster_level":true,"allow_gaps_topic_level":true}
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"nessie","cloud_storage_type":1}
tests/rptest/tests/archival_test.py::ArchivalTest.test_all_partitions_leadership_transfer@{"cloud_storage_type":1}
```
<!--/body-->

bashtanov commented on src/v/raft/tests/persisted_stm_test.cc:
```diff
@@ -766,3 +779,215 @@ TEST_F_CORO(state_machine_fixture, test_concurrent_apply_and_snapshot) {
+        add_node(id, model::revision_id(0), data_dir);
+    }
+    auto all_nodes = all_vnodes();
+    std::sort(all_nodes.begin(), all_nodes.end(), [](auto lhs, auto& rhs) {
+        return lhs.id() < rhs.id();
+    });
```
`std::ranges::sort(all_nodes, {}, &vnode::id)` ?

bashtanov commented on src/v/raft/tests/persisted_stm_test.cc:
```diff
@@ -766,3 +779,215 @@ TEST_F_CORO(state_machine_fixture, test_concurrent_apply_and_snapshot) {
+
+    for (const auto& [id, stm] : node_stms) {
+        ASSERT_EQ_CORO(stm->state, expected);
+        // snapshot was applied, not further applies should happen
```
typo not->no

bashtanov commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -185,10 +192,29 @@ ss::future<> state_machine_manager::start() {
+    /**
+     * Special case for the `archival_metadata_stm` local snapshot created
+     * during recovery. The snapshot last included offset was set to the log
+     * start offset.
```
Thanks for the more detailed comment, it made me realise I don't understand this special case. The condition below holds iff the log is empty. Why is archival_metadata_stm local snapshot special so it is not consistent with the log position?

mmaslankaprv replied:
yes, exactly

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64669-->
<!--d99b57f3-0c18-4527-a4b6-1babd354b581-->
<!--2ff77396-2cfe-4c76-9fe3-504f3a0c1523-->
<!--title-->
## Retry command for [Build#64669](https://buildkite.com/redpanda/redpanda/builds/64669)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all@{"use_transactions":false}
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
```
<!--/body-->

mmaslankaprv commented:
/ci-repeat 1

mmaslankaprv commented:
It looks like all those tests are just flaky and unrelated with this PR

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64687-->
<!--914e9ddd-3332-4679-a6cb-63e548f89a87-->
<!--88a9ead6-383e-4110-89c5-b0dc72f157c6-->
<!--title-->
## Retry command for [Build#64687](https://buildkite.com/redpanda/redpanda/builds/64687)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/node_pool_migration_test.py::NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"off","cleanup_policy":"compact","test_mode":"tiered_storage_fast_moves"}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/transactions/producers_api_test.py::ProducersAdminAPITest.test_producers_state_api_during_load
```
<!--/body-->

mmaslankaprv commented:
Force merging this PR to enable progress on the issue, the failures are unrelated and they are being worked on

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x
